### PR TITLE
Add functions `List.append_map` and `List.rev_append_map`.

### PR DESCRIPTION
**Pull request by SylvainBoilard**
Hi,

This is a tail-recursive-modulo-cons implementation of a combination of `map` and `append` in a single function, with a noticeable performance and memory usage improvement. I found this function useful in a web project when generating items for a single list with data extracted from different database tables.

<details>
<summary>Benchmark</summary>

[Benchmark source code](https://gist.github.com/SylvainBoilard/4ac2290189b1fc87745e78e1fca08e83)

For the unrolled thrice implementation proposed in this PR:

- twice as fast as the composition of `map` and `append` except for lists of 4 elements where the improvement is smaller ;
- faster than the composition of `rev_map` and `rev_append` except for lists of 4 elements, up to twice as fast for very long lists ;
- half as many allocations.

The unrolled thrice implementation is always faster than the naive version, and is a bit faster than the versions with less unrolls except for lists of 1, 4 and 8 elements.

| list length | composed | rev composed | specialized | specialized, unrolled once | specialized, unrolled twice | specialized, unrolled thrice |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
|           1 |     0.009 µs |     0.008 µs |    0.008 µs |      0.004 µs |      0.005 µs |      0.005 µs |
|             |          6 W |         11 W |         3 W |           3 W |           3 W |           3 W |
|           2 |     0.011 µs |     0.010 µs |    0.013 µs |      0.009 µs |      0.006 µs |      0.006 µs |
|             |         12 W |         17 W |         6 W |           6 W |           6 W |           6 W |
|           4 |     0.020 µs |     0.015 µs |    0.020 µs |      0.013 µs |      0.012 µs |      0.015 µs |
|             |         24 W |         29 W |        12 W |          12 W |          12 W |          12 W |
|           8 |     0.038 µs |     0.025 µs |    0.037 µs |      0.027 µs |      0.021 µs |      0.023 µs |
|             |         48 W |         53 W |        24 W |          24 W |          24 W |          24 W |
|          16 |     0.074 µs |     0.046 µs |    0.069 µs |      0.055 µs |      0.044 µs |      0.040 µs |
|             |         96 W |        101 W |        48 W |          48 W |          48 W |          48 W |
|          32 |     0.158 µs |     0.096 µs |    0.144 µs |      0.120 µs |      0.104 µs |      0.081 µs |
|             |        192 W |        197 W |        96 W |          96 W |          96 W |          96 W |
|          64 |     0.322 µs |     0.200 µs |    0.275 µs |      0.235 µs |      0.200 µs |      0.179 µs |
|             |        384 W |        389 W |       192 W |         192 W |         192 W |         192 W |
|         128 |     0.623 µs |     0.379 µs |    0.542 µs |      0.458 µs |      0.396 µs |      0.348 µs |
|             |        768 W |        773 W |       384 W |         384 W |         384 W |         384 W |
|         256 |     1.258 µs |     0.760 µs |    1.088 µs |      0.911 µs |      0.778 µs |      0.687 µs |
|             |       1536 W |       1541 W |       768 W |         768 W |         768 W |         768 W |
|         512 |     2.540 µs |     1.582 µs |    2.167 µs |      1.824 µs |      1.565 µs |      1.378 µs |
|             |       3072 W |       3077 W |      1536 W |        1536 W |        1536 W |        1536 W |
|        1024 |     5.400 µs |     3.313 µs |    4.414 µs |      3.717 µs |      3.191 µs |      2.792 µs |
|             |       6144 W |       6149 W |      3072 W |        3072 W |        3072 W |        3072 W |
|        2048 |    11.950 µs |     7.141 µs |    9.062 µs |      7.729 µs |      6.646 µs |      5.867 µs |
|             |      12288 W |      12293 W |      6144 W |        6144 W |        6144 W |        6144 W |
|        4096 |    28.896 µs |    16.187 µs |   19.280 µs |     16.816 µs |     14.578 µs |     13.145 µs |
|             |      24576 W |      24581 W |     12288 W |       12288 W |       12288 W |       12288 W |
|        8192 |    80.560 µs |    53.567 µs |   51.931 µs |     47.071 µs |     42.079 µs |     39.048 µs |
|             |      49152 W |      49157 W |     24576 W |       24576 W |       24576 W |       24576 W |
|       16384 |   253.816 µs |   199.398 µs |  150.614 µs |    141.389 µs |    140.099 µs |    131.142 µs |
|             |      98304 W |      98309 W |     49152 W |       49152 W |       49152 W |       49152 W |
|       32768 |   992.357 µs |   804.357 µs |  499.900 µs |    479.605 µs |    464.354 µs |    459.375 µs |
|             |     196608 W |     196613 W |     98304 W |       98304 W |       98304 W |       98304 W |
|       65536 |     3.591 ms |     2.686 ms |    1.579 ms |      1.544 ms |      1.519 ms |      1.517 ms |
|             |     393216 W |     393221 W |    196608 W |      196608 W |      196608 W |      196608 W |
|      131072 |     7.493 ms |     6.552 ms |    3.865 ms |      3.766 ms |      3.728 ms |      3.707 ms |
|             |     786432 W |     786437 W |    393216 W |      393216 W |      393216 W |      393216 W |
|      262144 |    15.078 ms |    13.795 ms |    7.773 ms |      7.634 ms |      7.553 ms |      7.513 ms |
|             |    1572864 W |    1572869 W |    786432 W |      786432 W |      786432 W |      786432 W |
|      524288 |    30.294 ms |    28.400 ms |   15.510 ms |     15.158 ms |     14.703 ms |     14.670 ms |
|             |    3145728 W |    3145733 W |   1572864 W |     1572864 W |     1572864 W |     1572864 W |
|     1048576 |    59.174 ms |    56.882 ms |   30.475 ms |     30.288 ms |     29.779 ms |     29.601 ms |
|             |    6291456 W |    6291461 W |   3145728 W |     3145728 W |     3145728 W |     3145728 W |
</details>

Please let me know of any issue.

Cheers,

### CODE REVIEW

**Comment by nojb** [comment]
We already have `List.concat_map` for an arbitrary number of lists (also TRMC). My feeling is that having a specialized version of this for the case of two lists is not worth it, but let's hear if there are other opinions.

**Comment by nojb** [comment]
(Thanks for including a detailed benchmark, by the way!)

**Comment by gasche** [comment]
For the record, the proposed function has the following signature:

```ocaml
val map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
```

**Comment by nojb** [comment]
> For the record, the proposed function has the following signature:

Ah, it is not the same as `List.concat_map`; I should have looked more carefully. Thanks for the correcting me.

**Comment by hhugo** [comment]
We have the rev version in bytecomp/bytepackager.ml

```ocaml
let rec rev_append_map f l rest =
  match l with
  | [] -> rest
  | x :: xs -> rev_append_map f xs (f x :: rest)
```

It exists in js_of_ocaml as well .

Should the rev version be added as well ?

**Comment by SylvainBoilard** [comment]
I went ahead and added the `rev` version pretty much as-is.

I also ran a benchmark with unrolled versions: for me it’s a mixed bag between the naive and the unrolled thrice versions, and I am under the impression that you usually go with the simpler approach in those cases. Let me know if I should pick another version (or none).

<details>
<summary>Benchmark</summary>

[Benchmark source code](https://gist.github.com/SylvainBoilard/4ac2290189b1fc87745e78e1fca08e83)

For the naive version implemented in this PR:
- Always faster than both composition options, about twice as fast with lists of more than 256 elements, about thrice as fast with lists between 8192 and 32768, and a bit more than twice as fast for longer lists ;
- half as many memory allocations.

The unrolled versions perform about the same as the naive version, although the unrolled twice version is a bit faster with lists up to 8 elements long, and the unrolled thrice version with lists up to 128 elements long (up to about 25% faster for lists of 2 and 8 elements, but otherwise only by a small amount). The unrolled thrice version appears to be sometimes faster, sometimes slower than the naive version with longer lists.

| list length | map_rev_app | rev_map_app | specialized | spec unroll×1 | spec unroll×2 | spec unroll×3 |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
|           1 |    0.007 µs |    0.007 µs |    0.005 µs |      0.005 µs |      0.005 µs |      0.005 µs |
|             |         6 W |        11 W |         3 W |           3 W |           3 W |           3 W |
|           2 |    0.016 µs |    0.009 µs |    0.008 µs |      0.009 µs |      0.006 µs |      0.006 µs |
|             |        12 W |        17 W |         6 W |           6 W |           6 W |           6 W |
|           4 |    0.021 µs |    0.019 µs |    0.011 µs |      0.013 µs |      0.010 µs |      0.010 µs |
|             |        24 W |        29 W |        12 W |          12 W |          12 W |          12 W |
|           8 |    0.039 µs |    0.030 µs |    0.019 µs |      0.021 µs |      0.017 µs |      0.016 µs |
|             |        48 W |        53 W |        24 W |          24 W |          24 W |          24 W |
|          16 |    0.073 µs |    0.054 µs |    0.031 µs |      0.034 µs |      0.031 µs |      0.029 µs |
|             |        96 W |       101 W |        48 W |          48 W |          48 W |          48 W |
|          32 |    0.156 µs |    0.116 µs |    0.065 µs |      0.075 µs |      0.070 µs |      0.064 µs |
|             |       192 W |       197 W |        96 W |          96 W |          96 W |          96 W |
|          64 |    0.311 µs |    0.224 µs |    0.121 µs |      0.144 µs |      0.136 µs |      0.119 µs |
|             |       384 W |       389 W |       192 W |         192 W |         192 W |         192 W |
|         128 |    0.620 µs |    0.428 µs |    0.233 µs |      0.285 µs |      0.263 µs |      0.232 µs |
|             |       768 W |       773 W |       384 W |         384 W |         384 W |         384 W |
|         256 |    1.235 µs |    0.850 µs |    0.457 µs |      0.563 µs |      0.497 µs |      0.466 µs |
|             |      1536 W |      1541 W |       768 W |         768 W |         768 W |         768 W |
|         512 |    2.458 µs |    1.812 µs |    0.961 µs |      1.168 µs |      0.996 µs |      0.927 µs |
|             |      3072 W |      3077 W |      1536 W |        1536 W |        1536 W |        1536 W |
|        1024 |    5.345 µs |    3.626 µs |    1.897 µs |      2.314 µs |      2.112 µs |      1.937 µs |
|             |      6144 W |      6149 W |      3072 W |        3072 W |        3072 W |        3072 W |
|        2048 |   12.085 µs |    8.067 µs |    4.083 µs |      4.852 µs |      4.621 µs |      4.026 µs |
|             |     12288 W |     12293 W |      6144 W |        6144 W |        6144 W |        6144 W |
|        4096 |   27.729 µs |   20.107 µs |    9.199 µs |     10.730 µs |      9.630 µs |      9.228 µs |
|             |     24576 W |     24581 W |     12288 W |       12288 W |       12288 W |       12288 W |
|        8192 |   74.979 µs |   76.460 µs |   26.876 µs |     30.232 µs |     27.519 µs |     26.770 µs |
|             |     49152 W |     49157 W |     24576 W |       24576 W |       24576 W |       24576 W |
|       16384 |  261.284 µs |  254.945 µs |   96.378 µs |     94.110 µs |     86.558 µs |     87.051 µs |
|             |     98304 W |     98309 W |     49152 W |       49152 W |       49152 W |       49152 W |
|       32768 |  981.525 µs |  937.675 µs |  322.519 µs |    330.944 µs |    326.272 µs |    323.579 µs |
|             |    196608 W |    196613 W |     98304 W |       98304 W |       98304 W |       98304 W |
|       65536 |    3.548 ms |    3.307 ms |    1.356 ms |      1.381 ms |      1.364 ms |      1.353 ms |
|             |    393216 W |    393221 W |    196608 W |      196608 W |      196608 W |      196608 W |
|      131072 |    7.748 ms |    7.502 ms |    3.610 ms |      3.621 ms |      3.589 ms |      3.612 ms |
|             |    786432 W |    786437 W |    393216 W |      393216 W |      393216 W |      393216 W |
|      262144 |   16.078 ms |   15.843 ms |    7.620 ms |      7.935 ms |      7.904 ms |      7.864 ms |
|             |   1572864 W |   1572869 W |    786432 W |      786432 W |      786432 W |      786432 W |
|      524288 |   33.644 ms |   33.366 ms |   16.261 ms |     16.129 ms |     15.996 ms |     16.169 ms |
|             |   3145728 W |   3145733 W |   1572864 W |     1572864 W |     1572864 W |     1572864 W |
|     1048576 |   66.584 ms |   64.959 ms |   31.754 ms |     32.026 ms |     31.852 ms |     31.841 ms |
|             |   6291456 W |   6291461 W |   3145728 W |     3145728 W |     3145728 W |     3145728 W |
</details>

File: stdlib/list.mli [hunk 1]
```diff
@@ -192,6 +192,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
    {!rev}[ (]{!map}[ f l)], but is more efficient.
  *)
 
+val map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [map_append f lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ f lm @ la], but is more efficient.
+    @since 5.03
```
Comment (nojb) [inline stdlib/list.mli]:
```suggestion
    @since 5.4
```

File: stdlib/list.mli [hunk 2]
```diff
@@ -192,6 +192,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
...
+ *)
+
+val rev_map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [rev_map_append f lm la] applies [f] to every element of [lm] and
+    appends the result in reverse order to [la]. This is equivalent to
+    {!rev_map}[ f lm @ la], but is more efficient.
+    @since 5.03
```
Comment (nojb) [inline stdlib/list.mli]:
```suggestion
    @since 5.4
```

File: stdlib/listLabels.mli [hunk 3]
```diff
@@ -192,6 +192,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
    {!rev}[ (]{!map}[ f l)], but is more efficient.
  *)
 
+val map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
+(** [map_append ~f ~lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ ~f lm @ la], but is more efficient.
+    @since 5.03
```
Comment (nojb) [inline stdlib/listLabels.mli]:
```suggestion
    @since 5.4
```

File: stdlib/listLabels.mli [hunk 4]
```diff
@@ -192,6 +192,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
...
+ *)
+
+val rev_map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
+(** [rev_map_append ~f ~lm la] applies [f] to every element of [lm] and
+    appends the result in reverse order to [la]. This is equivalent to
+    {!rev_map}[ ~f lm @ la], but is more efficient.
+    @since 5.03
```
Comment (nojb) [inline stdlib/listLabels.mli]:
```suggestion
    @since 5.4
```

**Review by nojb — APPROVED** [review]
The proposal seems reasonable to me. However, a second approval from a core dev is still necessary.

Also, a Changes entry is in order.

LGTM

**Comment by SylvainBoilard** [comment]
I updated the `testsuite/tests/backtrace/backtrace_dynlink.flambda.reference` backtrace that caused an error in the CI build.

There are also backtraces referencing the List module that might need updating in `testsuite/tests/statmemprof/callstacks.{,no-}flat-float-array.reference`, should I update these as well? If so, how can I run the corresponding tests on my side?

**Comment by nojb** [comment]
Friendly ping @damiendoligez.

File: stdlib/list.mli [hunk 5]
```diff
@@ -192,6 +192,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
    {!rev}[ (]{!map}[ f l)], but is more efficient.
  *)
 
+val map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [map_append f lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ f lm @ la], but is more efficient.
+    @since 5.4
```
Comment (nojb) [inline stdlib/list.mli]:
```suggestion
    @since 5.5
```

File: stdlib/list.mli [hunk 6]
```diff
@@ -192,6 +192,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
...
+ *)
+
+val rev_map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [rev_map_append f lm la] applies [f] to every element of [lm] and
+    appends the result in reverse order to [la]. This is equivalent to
+    {!rev_map}[ f lm @ la], but is more efficient.
+    @since 5.4
```
Comment (nojb) [inline stdlib/list.mli]:
```suggestion
    @since 5.5
```

File: stdlib/listLabels.mli [hunk 7]
```diff
@@ -192,6 +192,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
    {!rev}[ (]{!map}[ f l)], but is more efficient.
  *)
 
+val map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
+(** [map_append ~f ~lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ ~f lm @ la], but is more efficient.
+    @since 5.4
```
Comment (nojb) [inline stdlib/listLabels.mli]:
```suggestion
    @since 5.5
```

File: stdlib/listLabels.mli [hunk 8]
```diff
@@ -192,6 +192,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
...
+ *)
+
+val rev_map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
+(** [rev_map_append ~f ~lm la] applies [f] to every element of [lm] and
+    appends the result in reverse order to [la]. This is equivalent to
+    {!rev_map}[ ~f lm @ la], but is more efficient.
+    @since 5.4
```
Comment (nojb) [inline stdlib/listLabels.mli]:
```suggestion
    @since 5.5
```

File: stdlib/listLabels.mli [hunk 9]
```diff
@@ -192,6 +192,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
...
  *)
 
+val map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
+(** [map_append ~f ~lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ ~f lm @ la], but is more efficient.
+    @since 5.4
+ *)
+
+val rev_map_append : f:('a -> 'b) -> lm:'a list -> 'b list -> 'b list
```
Comment (nojb) [inline stdlib/listLabels.mli]: I am not sure about the `~lm` label: it doesn't stand for anything. I think it would be better to remove it (or find a better name).

**Review by nojb — APPROVED** [review]
LGTM, modulo the `~lm` label (I am re-reviewing because a lot of time has passed).

Can we get another review of this PR? Maybe @OlivierNicole or @Octachron?

File: stdlib/listLabels.mli (hunk 9 above)
Comment (SylvainBoilard) [reply stdlib/listLabels.mli]: It is supposed to refer to the list to be mapped, perhaps a more verbose label would be preferable?

Comment (nojb) [reply stdlib/listLabels.mli]: Personally, I don't think a label is needed because the types tell the whole story; there is only one of the lists that can be mapped over.

Comment (SylvainBoilard) [reply stdlib/listLabels.mli]: I think I misunderstood the purpose of the `ListLabels` module, as I thought it was supposed to provide users with the possibility of partially applying its functions to any of their parameters, but in reality some of the functions in there do have fewer labels than I expected. Could you clarify this point?

Comment (nojb) [reply stdlib/listLabels.mli]: That is one of its purposes, but the idea is not necessarily to allow commuting _any_ argument, but typically those which are more common and/or useful (here, the functional argument). There are many existing three-argument functions in this module with only one labelled functional argument (eg `map2`). So we could follow this existing style here.

**Comment by dbuenzli** [comment]
Not sure how many more of these pre-composed functions we are going to add to the `List` module but I think the proposed names are *utterly* confusing.

```ocaml
List.map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
List.rev_map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
``` 

I honestly have no idea how to interpret this:

```ocaml
let l = List.rev_map_append f l0 l1 
```

Based on [previous](https://github.com/ocaml/ocaml/pull/14185#issuecomment-3173775094) experience with `List` module I would infer that this append `l0` and `l1`, maps it with `f` and then reverses the list. That's not what it does though. 

If the proposal was a minimal consistent with itself it should rather be: 

```ocaml
List.map_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
List.map_rev_append : ('a -> 'b) -> 'a list -> 'b list -> 'b list
```
But you'd still have to convince me that this doesn't read like `[rev_]append` the lists and then `map` them.
```ocaml
let l = List.map_append f l0 l1
let l = List.map_rev_append f l0 l1 
```

Now if we were to follow the current `List` module scheme that would rather be: 

```ocaml
List.append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
List.rev_append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
```
At least we can now apply our previous knowledge of the `List` module to reason on what these functions could be doing, but on reading you still have to convince your brain not associate the `_map` as being the object of the `append` or being applied to the second list.

```ocaml
let l = List.append_map f l0 l1
let l = List.rev_append_map f l0 l1 
```

Perhaps not having these functions is better and people who need the performance gain can define them in their project ?

```ocaml 
let l = List.append (List.map f l0) l1
let l = List.rev_append (List.map f l0) l1
```

**Comment by SylvainBoilard** [comment]
Thank you for you feedback, I think renaming these functions `append_map` and `append_rev_map` would in fact make more sense from a function composition point of view, as well as for consistency.

```ocaml
List.append_map f l0 l1 = List.(append (map f l0) l1)
List.append_rev_map f l0 l1 = List.(append (rev (map f l0)) l1)
```

File: stdlib/listLabels.mli (hunk 9 above)
Comment (SylvainBoilard) [reply stdlib/listLabels.mli]: Thank you for your feedback, I will remove the label then.

**Comment by dbuenzli** [comment]
> `append_rev_map` would in fact make more sense from a function composition point of view, as well as for consistency.

No, I don't think `append_rev_map` is good a name. You are not doing: 

```ocaml
List.append_rev_map f l0 l1 = List.append (List.rev (List.map f l0)) l1
```
you are doing: 
```ocaml
List.rev_append_map f l0 l1 = List.rev_append (List.map f l0) l1
```

For anyone acquainted with the `List` module `List.rev_append` is an established operation with better properties than `List.append (List.rev …)` so I rather have that in the name.

**Comment by gasche** [comment]
I agree that `append_map` and `rev_append_map` are reasonable names.

File: stdlib/list.mli [hunk 10]
```diff
@@ -197,6 +197,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
    {!rev}[ (]{!map}[ f l)], but is more efficient.
  *)
 
+val append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [append_map f lm la] applies [f] to every element of [lm] and
+    appends the result to [la]. This is equivalent to
+    {!map}[ f lm @ la], but is more efficient.
```
Comment (dbuenzli) [inline stdlib/list.mli]: I think you should rather use the equations we discussed instead of `@`, the name of the function is literally a contraction of the expression they stand for so it will make it easier for users to eventually engrave the convention in their head.

Also the rest of the module uses `l1` and `l2` as variable for `append` & co. I find these `lm` and `la` distracting (because the acronym expansion part of my brain wants to find out a meaning to them but it's not useful to do so, the equation is clear enough :-). 

Finally you don't append `(map l1)` to `l2`, it's the converse and that's going to be `@since 5.6`.

TLDR, I think this would be better a better doc string:

```ocaml
(** [append_map f l1 l2] applies [f] to every element of [l1] and appends
    [l2] to the result. This is {!append}[ (]{!map}[ f l1) l2] but more
    efficient. 

    @since 5.6 *)
```

File: stdlib/list.mli [hunk 11]
```diff
@@ -197,6 +197,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
...
+    @since 5.5
+ *)
+
+val rev_append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [rev_append_map f lm la] applies [f] to every element of [lm] and
+    appends the result in reverse order to [la]. This is equivalent to
+    {!rev_map}[ f lm @ la], but is more efficient.
+    @since 5.5
+ *)
```
Comment (dbuenzli) [inline stdlib/list.mli:212]: Same comments here

```ocaml
(** [rev_append_map f l1 l2] applies [f] to every element of [l1], reverse the
    result and appends [l2] to it. This is {!rev_append}[ (]{!map}[ f l1) l2] 
    but more efficient.

    @since 5.6 *)

**Comment by nojb** [comment]
This PR seems to be in a mergeable state.

It only needs a second approval from a core maintainer (NB: I am not submitting this to the stdlib team because PR has already been thoroughly discussed and reworked and seems to be in a consensual state).

Maybe @gasche can do a final review?

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
I agree that the functions can be useful, and reviewed the implementation for correctness.

File: stdlib/list.ml [hunk 12]
```diff
@@ -108,6 +108,17 @@ let rev_map f l =
   in
   rmap_f [] l
 
+let[@tail_mod_cons] rec append_map f l1 l2 = match l1 with
+  | [] -> l2
+  | [ hd ] -> f hd :: l2
+  | [ hd1 ; hd2 ] -> f hd1 :: f hd2 :: l2
+  | [ hd1 ; hd2 ; hd3 ] -> f hd1 :: f hd2 :: f hd3 :: l2
+  | hd1 :: hd2 :: hd3 :: hd4 :: tl ->
+     f hd1 :: f hd2 :: f hd3 :: f hd4 :: append_map f tl l2
```
Comment (yallop) [inline stdlib/list.ml]: The function should let-bind the expressions that produce the intermediate results to avoid the surprising behaviour of the current implementation:

```
# append_map (Printf.printf "%d\n") (List.init 10 Fun.id) [];;
3
4
5
6
7
9
8
2
1
0
```

**Review by yallop — CHANGES_REQUESTED** [review]

File: stdlib/list.ml (hunk 12 above)
Comment (nojb) [reply stdlib/list.ml]: Oops, good catch. This could be added as a test as well.

Comment (nojb) [reply stdlib/list.ml]: I took the liberty of pushing these changes and adding @yallop to the list of reviewers.

@yallop: can you review the changes and approve formally, if you agree with them? Thanks!

File: stdlib/list.mli [hunk 13]
```diff
@@ -197,6 +197,20 @@ val rev_map : ('a -> 'b) -> 'a list -> 'b list
...
 
+val append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [append_map f l1 l2] applies [f] to every element of [l1] and appends
+    [l2] to the result. This is {!append}[ (]{!map}[ f l1) l2] but more
+    efficient.
+    @since 5.6
+ *)
+
+val rev_append_map : ('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [rev_append_map f l1 l2] applies [f] to every element of [l1], reverse the
```
Comment (yallop) [inline stdlib/list.mli]:
```suggestion
(** [rev_append_map f l1 l2] applies [f] to every element of [l1], reverses the
```

File: stdlib/listLabels.mli [hunk 14]
```diff
@@ -197,6 +197,20 @@ val rev_map : f:('a -> 'b) -> 'a list -> 'b list
...
 
+val append_map : f:('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [append_map ~f l1 l2] applies [f] to every element of [l1] and appends
+    [l2] to the result. This is {!append}[ (]{!map}[ f l1) l2] but more
+    efficient.
+    @since 5.6
+ *)
+
+val rev_append_map : f:('a -> 'b) -> 'a list -> 'b list -> 'b list
+(** [rev_append_map ~f l1 l2] applies [f] to every element of [l1], reverse the
```
Comment (yallop) [inline stdlib/listLabels.mli]:
```suggestion
(** [rev_append_map ~f l1 l2] applies [f] to every element of [l1], reverses the
```

**Comment by gasche** [comment]
Apologies for missing the evaluation-order issue, I was evidently too hasty in my review. Thanks @yallop for being more careful!

File: stdlib/listLabels.mli (hunk 14 above)
Comment (nojb) [reply stdlib/listLabels.mli]: Applied, thanks.

File: stdlib/list.mli (hunk 13 above)
Comment (nojb) [reply stdlib/list.mli]: Applied, thanks.

**Review by yallop — APPROVED** [review]

**Comment by SylvainBoilard** [comment]
Thanks to all involved for the time spent on reviewing, improving and merging this PR!

Cheers.
